src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,304 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QDebug>
+
+#include <Qt3DRender/QObjectPicker>
+#include <Qt3DRender/QPickEvent>
+#include <QKeyEvent>
+
+#include "qgsterrainentity_p.h"
+#include "qgs3dmaptoolmeasureline.h"
+#include "qgs3dutils.h"
+#include "qgs3dmapscene.h"
+#include "qgs3dmapcanvas.h"
+#include "qgsvectorlayer.h"
+#include "qgslinestring.h"
+#include "qgspoint.h"
+#include "qgsfeature.h"
+#include "qgsline3dsymbol.h"
+#include "qgsvectorlayer3drenderer.h"
+#include "qgsmaplayer.h"
+#include "qgs3dmeasuredialog.h"
+
+#include "qgs3dmapscenepickhandler.h"
+
+class Qgs3DMapToolMeasureLinePickHandler : public Qgs3DMapScenePickHandler
+{
+  public:
+    Qgs3DMapToolMeasureLinePickHandler( Qgs3DMapToolMeasureLine *measureLineTool ): mMeasureLineTool( measureLineTool ) {}
+    void handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event ) override;
+  private:
+    Qgs3DMapToolMeasureLine *mMeasureLineTool = nullptr;
+};
+
+void Qgs3DMapToolMeasureLinePickHandler::handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event )
+{
+  if ( event->button() == Qt3DRender::QPickEvent::LeftButton )
+  {
+    if ( mMeasureLineTool->done() )
+    {
+      mMeasureLineTool->mDialog->restart();
+    }
+
+    // Left button, keep addinng point
+    qInfo() << "handlePickOnVectorLayer" ;
+    QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates(
+                              QgsVector3D( worldIntersection.x(),
+                                           worldIntersection.y(),
+                                           worldIntersection.z() ), mMeasureLineTool->mCanvas->map()->origin() );
+    QgsPoint pt( mapCoords.x(), mapCoords.y(), mapCoords.z() );
+    qInfo() << "Coord (handlePickOnVectorLayer): " << pt.x() << " " << pt.y() << " " << pt.z();
+
+    mMeasureLineTool->mDone = false;
+    mMeasureLineTool->addPoint( pt );
+    mMeasureLineTool->mDialog->show();
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::RightButton )
+  {
+    // Finish measurement
+    qInfo() << "Finish measurement";
+    mMeasureLineTool->mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    qInfo() << "Undo - middle button";
+    mMeasureLineTool->undo();
+
+  }
+}
+
+Qgs3DMapToolMeasureLine::Qgs3DMapToolMeasureLine( Qgs3DMapCanvas *canvas )
+  : Qgs3DMapTool( canvas )
+{
+  qInfo() << "Constructed";
+  connect( mCanvas->scene(), &Qgs3DMapScene::terrainEntityChanged, this, &Qgs3DMapToolMeasureLine::onTerrainEntityChanged );
+  mPickHandler.reset( new Qgs3DMapToolMeasureLinePickHandler( this ) );
+
+  // Line style
+  mLineSymbol = new QgsLine3DSymbol;
+  mLineSymbol->setRenderAsSimpleLines( true );
+  mLineSymbol->setWidth( 4 );
```

Comment:
As discussed in our chat today, would be good to use Absolute altitude clamping so that the terrain is not added twice.

Reply:
Done.